core/src/server/rpc_module.rs

```diff
@@ -775,69 +775,50 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	}
 }
 
-/// Represent a pending subscription which waits to be accepted or rejected.
-///
-/// Note: you need to call either `PendingSubscription::accept` or `PendingSubscription::reject` otherwise
-/// the subscription will be dropped with an `InvalidParams` error.
+/// The state of the [`SubscriptionSink`].
 #[derive(Debug)]
-struct InnerPendingSubscription {
-	/// Sink.
-	sink: MethodSink,
-	/// Get notified when subscribers leave so we can exit
-	close_notify: Option<SubscriptionPermit>,
-	/// MethodCallback.
-	method: &'static str,
-	/// Unique subscription.
-	uniq_sub: SubscriptionKey,
-	/// Shared Mutex of subscriptions
-	subscribers: Subscribers,
-	/// Request ID.
-	id: Id<'static>,
-	/// Claimed resources.
-	claimed: Option<ResourceGuard>,
+enum SubscriptionSinkState {
+	/// The subscription is pending and needs to be accepted or rejected
+	/// before utilizing the sink. This is the initial state of the sink.
+	///
+	/// This state contains the request ID of the subscription.
+	Pending(Option<Id<'static>>),
+	/// The subscription was accepted via the [`SubscriptionSink::accept`] call.
+	/// This state contains the future that returns when the unsubscribe method has been called.
```

Comment:
I'm still struggling to understand the `Accept which holds` the `rx` which yields once the unsubscribe method has been called.

It just makes me confused as we have `method_sink` and `close_notify` which is sort of something similar which is in the `SubscriptionSink` it just makes everything hard to follow.

Thus, I'm in favor of having everything in the `SubscriptionSink` and just remove `SubscriptionSinkState` because the method sink and close_notify are technically part of that state too.

Alternatively write some nicer abstraction for that.

Thoughts?

Reply:
Sounds like removing it might make our life easier, tho we'd have to be careful internally about how to reason about the states in which we are in.
I could remodel it using `Option<Id<'static>>` for "pending", `Option<watch::Receiver<()>` for "accepted" and treat other cases as "rejected" via the `close_notify` option.

Reply:
yeah, having different types for the states was one option :D